pkg/binlog/binlog.go

```diff
@@ -0,0 +1,294 @@
+func (c *Client) IsPumpTombstone(ctx context.Context, addr string) (bool, error) {
+	nodeID, err := c.nodeID(ctx, addr, "pumps")
+	if err != nil {
+		return false, err
+	}
+	return c.isTombstone(ctx, "pumps", nodeID)
+}
+
+// IsDrainerTombstone check if drainer is tombstone.
+func (c *Client) IsDrainerTombstone(ctx context.Context, addr string) (bool, error) {
+	nodeID, err := c.nodeID(ctx, addr, "drainers")
+	if err != nil {
+		return false, err
+	}
+	return c.isTombstone(ctx, "drainer", nodeID)
+}
+
+func (c *Client) isTombstone(ctx context.Context, ty string, nodeID string) (bool, error) {
+	status, err := c.nodeStatus(ctx, ty)
+	if err != nil {
+		return false, err
+	}
+
+	for _, s := range status {
+		if s.NodeID == nodeID {
+			if s.State == "offline" {
+				return true, nil
+			}
+			return false, nil
+		}
+	}
+
+	return false, errors.Errorf("node not exist: %s", nodeID)
+}
+
+func (c *Client) PumpNodeStatus(ctx context.Context) (status []*NodeStatus, err error) {
+	return c.nodeStatus(ctx, "pumps")
+}
+
+// nolint (unused)
+func (c *Client) drainerNodeStatus(ctx context.Context) (status []*NodeStatus, err error) {
+	return c.nodeStatus(ctx, "drainers")
+}
+
+func (c *Client) nodeID(ctx context.Context, addr, ty string) (string, error) {
+	nodes, err := c.nodeStatus(ctx, ty)
+	if err != nil {
+		return "", err
+	}
+
+	addrs := []string{}
+	for _, node := range nodes {
+		if addr == node.Addr {
+			return node.NodeID, nil
+		}
+		addrs = append(addrs, addr)
+	}
+
+	return "", errors.Errorf("%s node id for address %s not found, found address: %s", ty, addr, addrs)
+}
+
+// UpdateDrainerState update the specify state as the specified state.
+func (c *Client) UpdateDrainerState(ctx context.Context, addr string, state string) error {
+	nodeID, err := c.nodeID(ctx, addr, "drainers")
+	if err != nil {
+		return err
+	}
+	return c.updateStatus(ctx, "drainers", nodeID, state)
+}
+
+// UpdatePumpState update the specify state as the specified state.
+func (c *Client) UpdatePumpState(ctx context.Context, addr string, state string) error {
+	nodeID, err := c.nodeID(ctx, addr, "pumps")
+	if err != nil {
+		return err
+	}
+	return c.updateStatus(ctx, "pumps", nodeID, state)
+}
+
+// updateStatus update the specify state as the specified state.
+func (c *Client) updateStatus(ctx context.Context, ty string, nodeID string, state string) error {
+	key := fmt.Sprintf("/tidb-binlog/v1/%s/%s", ty, nodeID)
+
+	resp, err := c.etcdClient.KV.Get(ctx, key)
+	if err != nil {
+		return errors.AddStack(err)
+	}
+
+	if len(resp.Kvs) == 0 {
+		return errors.Errorf("no pump with node id: %v", nodeID)
+	}
+
+	var nodeStatus NodeStatus
+	err = json.Unmarshal(resp.Kvs[0].Value, &nodeStatus)
+	if err != nil {
+		return errors.AddStack(err)
+	}
+
+	if nodeStatus.State == state {
+		return nil
+	}
+
+	nodeStatus.State = state
+
+	data, err := json.Marshal(&nodeStatus)
+	if err != nil {
+		return errors.AddStack(err)
+	}
+
+	_, err = c.etcdClient.Put(ctx, key, string(data))
```

Comment:
Maybe we need to compare (with the old value) and update **later** for concurrency safety?

Reply:
no